pkgs/development/interpreters/python/mk-python-derivation.nix

```
@@ -104,6 +104,49 @@ else
 let
   inherit (python) stdenv;
 
+  validatePythonMatches = attrName: let
+    isPythonModule = drv:
+      # all pythonModules have the pythonModule attribute
+      (drv ? "pythonModule")
+      # Some pythonModules are turned in to a pythonApplication by setting the field to false
+      && (!builtins.isBool drv.pythonModule);
+    isMismatchedPython = drv: drv.pythonModule != python;
+
+    optionalLocation = let
+        pos = builtins.unsafeGetAttrPos (if attrs ? "pname" then "pname" else "name") attrs;
+      in if pos == null then "" else " in ${pos.file}:${toString pos.line}:${toString pos.column}";
+
+    leftPadName = name: against: let
+        len = lib.max (lib.stringLength name) (lib.stringLength against);
+      in lib.strings.fixedWidthString len " " name;
+
+    drvName = drv: drv.pname or drv.name;
+    warnMismatch = drv: lib.warn ''
+      Warning: Python version mismatch in '${name}':
```

`Warning:` seems unnecessary; `lib.warn` automatically prepends `warning:` (and `trace:`, but...).

```suggestion
      Python version mismatch in '${name}' detected:
```